Python/chapter01/1.5 - OneAway/miguel_1.5_sol.py

```diff
@@ -18,27 +17,43 @@ def one_away(s1: str, s2: str) -> bool:
     2. remove a character
     3. replace a character
     Zero edits away means the two strings are exactly the same.
-    We will analyze character frequencies to detect anomalies.
-    An anomaly can be any of the 3 types of edits.
-    Runtime:  O(n + m), where n is length of s1 and m is length of s2
-    Space Complexity:  O(n + m)
+    We will do different checks depending on the size different of the strings.
+    Runtime:  O(n), where n is length of s1
+    Space Complexity:  O(n + n +/- 1) = O(n), where n is length of s1
     :param s1: one string we want to compare
     :param s2: other string to compare
     :return: true if one or zero edits away, false otherwise
     """
-    # if size differs by more than 1, guaranteed more than one anomaly.
+    # if size differs by more than 1, guaranteed more than one edit distance.
     if abs(len(s1) - len(s2)) > 1:
         return False
+    # if size is the same, scan for edits
+    if len(s1) == len(s2):
+        edits = 0
+        for i, c in enumerate(s1):
+            if c == s2[i]:
+                continue
+            edits += 1
+            if edits > 1:
+                return False
+        return True
+    # otherwise, we have an insertion/deletion
+    # compare both strings for insertion/deletion
+    # want to loop through the shorter string to avoid going out of bounds
+    s_short = s1 if len(s1) < len(s2) else s2
+    s_long = s1 if len(s1) > len(s2) else s2
 
-    char_freqs_s1 = collections.Counter(s1)
-    char_freqs_s2 = collections.Counter(s2)
-
-    anomalies = 0
-    for key, value in char_freqs_s1.items():
-        if key not in char_freqs_s2 or value != char_freqs_s2[key]:
-            anomalies += 1
-        if anomalies > 1:
-            return False
+    addend = 0
+    for i, c in enumerate(s_short):
+        if c == s_long[i + addend]:
+            continue
+        if i == i + addend:
```

Comment:
this could be simplified to `if not added:`.

but then again, it might be even better to keep consistency with L36-38:

```python
added += 1
if added > 1:
  return False
```